📚 Break down the status effect state machine into smaller steps

tomrijnbeek opened the pull request:
## ✨ What's this?
This PR breaks the existing logic for elemental status effects into smaller steps that different scopes can hook into.

## 🔍 Why do we want this?
For multiple status effects we were doing a bit of a hacky approach where we were reapplying their upgrade every tick. With this approach, there are now multiple points where the scopes can hook into changes to the active effect so they can only update certain things when needed.

## 🏗 How is it done?
The lifecycle of status effects is now as follows:

| Method | Called when | Used for | Returns |
|--------|--------|--------|--------|
| `StartScope` | When the scope becomes active because any effect will become active | Initialise common elements such as visual components that do not rely on effect instance | Can set an `out ElementalStatus` to apply a status icon |
| `StartEffect` | Whenever a new effect in this scope becomes active (may be multiple times per `StartScope`) | Apply elements specific to the current effect instance (e.g. apply an upgrade) | Can call `ChangeStatus` on the passed in context object to change the status icon to something new |
| `ApplyEffectTick` | Each tick as long as the scope is active | Apply recurring effects (e.g. damage over time) | - |
| `EndEffect` | Whenever an active effect stops being active (a new effect may start immediately with `StartEffect`) | Clean up all elements from `StartEffect`, including undoing their effects if needed | - |
| `EndScope` | When the scope becomes inactive because all active effects expired | Clean up all elements from `StartScope` | - |

As mentioned in the table above, I passed in a context object in `StartEffect` that allows modifying the status. Perhaps it can be used for other things as well. Taking this approach a step further, we could even have the context object automatically be responsible for cleanup (e.g. `context.ApplyUpgrade(upgrade)` could automatically undo that upgrade when the effect ends). For now, I decided to stick to the bare minimal requirements to see if this approach is worth pursuing at all.

## 💡 Review hints
1. I'm not particularly happy with the naming of the lifecycle methods. Suggestions for better names welcome

tomrijnbeek committed:
🔀 Always start effect before applying effect tick

tomrijnbeek committed:
📚 Break down the status effect state machine

tomrijnbeek committed:
🔤 Rename lifecycle methods

paulcscharf commented on src/Bearded.TD/Game/Simulation/Elements/ElementalPhenomenonScopeBase.cs:
```diff
@@ -46,21 +46,46 @@ public void ApplyTick(Instant now)
+
+        // effect A -> effect B
+        var statusIcon = activeEffect.StatusIcon;
+        EndEffect(target, effect);
+        startEffect(effect, out var createNewStatus);
```
I think I'd prefer a single method that SwitchEffect or similarly named method, since that may require less or different work (or no work at all) compared to actual stop/start. Also, right now this looks like a bug that it passes in the same `effect` twice?

paulcscharf replied:
Also, right now we allow leaving the status alone, or creating a new one, feels like we should allow unsetting it too?

tomrijnbeek replied:
Yep, this is indeed a bug, good catch.

I disagree with making a single `SwitchEffect` though. I went back and forth multiple times when writing this and decided against this. Why? Because each scope itself would just implement it by calling `EndEffect` and `StartEffect` in sequence. _Maybe_ they can be smart if they have an existing component by just tweaking its state, but we don't do that anywhere right now.

Scopes will still want to have some operations they do only on `StartScope` and `EndScope` (like adding the fire flicker component, to make sure we don't just readd that every tick), so essentially they will be building the exact same internal structure, where they have a `startEffect` and `endEffect` method that they call from both `StartScope` and `EndScope` respectively, as well as in `SwitchEffect`. So if this is common logic to all components, why not handle it in the base class to begin with?

paulcscharf commented on src/Bearded.TD/Game/Simulation/Elements/ElementalPhenomenonScopeBase.cs:
```diff
@@ -46,21 +46,46 @@ public void ApplyTick(Instant now)
+        // null -> effect A
+        if (activeEffect is null)
+        {
+            StartScope(target, out var createStatus);
+            startEffect(effect, out var createOverrideStatus);
```
Do we really need two separate methods if we always call both together anyways? Feels like it could just be a single one to simplify things.

tomrijnbeek replied:
In line 68, `startEffect` is called without `StartScope` before it.

paulcscharf commented on src/Bearded.TD/Game/Simulation/Elements/ElementalPhenomenonScopeBase.cs:
```diff
@@ -46,21 +46,46 @@ public void ApplyTick(Instant now)
+        }
+
+        // effect A -> effect A
+        if (activeEffect.Effect.Equals(effect)) return;
```
Nothing guarantees a sensible equality implementation - could this cause problems?
Instead, should we perhaps have a dirty flag that can be set?

tomrijnbeek replied:
Honestly, this could just be a `RefEquals`, but we can't use that because structs. We just want to know whether we are talking about a different effect then before. If it has the exact same parameters, that'd be funny but realistically shouldn't happen. I don't know if I would want to work with a dirty flag, since especially enemies might be hit lots of times.

Imagine an enemy being shocked big time by a conductive sniper hit, and then being hit by a million gatling hits trying to shock it a tiny bit. The sniper effect will continue to be the strongest (because it scales with damage), but if we work with a dirty flag, we'll be reapplying that same effect basically each tick because of the gatlings. Now, if the sniper hit was never there we'd still be applying a new effect each frame, so it's not like we can't prevent it, but to me the underlying idea doesn't feel right.

Not willing to die on this hill though so if you still think that's what we should do, can do.

paulcscharf committed:
🤭 Further refactor of element system

- Minimizes allocations
- Minimizes boilerplate
- Simplifies lifecycle slightly

tomrijnbeek committed:
🐛 Fix bug where EndEffect was called with new effect

paulcscharf committed:
🚧

tomrijnbeek committed:
Merge branch 'better-status-effects-2' into better-status-effects
